File: functions/src/calculations.ts
/**
 * Shared signal calculations used by both the Performance Model and the
 * Risk Model (Section 13.3 — "not fed separately, same raw data").
 *
 * Calendar windows (not check-in counts): 7-day / 28-day load, recovery
 * comparison, and persistent-fatigue look at real UTC calendar days ending
 * on `asOf`. A missed day is 0 training load. It is NOT skipped, so an
 * athlete who missed 3 days does not have their "7-day load" stretch across
 * 10 calendar days.
 *
 * Wellness averages (sleep, HRV, resting HR, fatigue): missing days and
 * null fields are excluded from the mean. They are never treated as 0 —
 * 0 hours of sleep is not a real logged value.
 *
 * All research-verified formulas are cited inline (Section 18.1):
 *  - Training Load = Duration × RPE (session-RPE method)
 *  - ACWR = 7-day acute load ÷ 28-day chronic average weekly load
 *  - Fitness/Fatigue = Banister impulse-response model
 */

export interface DailyEntry {
  date: string; // yyyy-mm-dd
  trainingLoad: number | null; // 0 on a logged rest day; missing day also sums as 0
  sleepHours: number | null;
  restingHeartRate: number | null;
  hrv: number | null;
  fatigueScore: number; // 1-5 when a check-in exists
  sessionSport?: string | null;
  sessionSportGroup?: string | null;
}

export interface CalendarDay {
  date: string;
  entry: DailyEntry | null;
}

function pad2(n: number): string {
  return String(n).padStart(2, '0');
}

/** UTC yyyy-mm-dd. */
export function utcDateKey(d = new Date()): string {
  return `${d.getUTCFullYear()}-${pad2(d.getUTCMonth() + 1)}-${pad2(d.getUTCDate())}`;
}

export function shiftDateKey(dateKey: string, deltaDays: number): string {
  const [y, m, day] = dateKey.split('-').map(Number);
  const dt = new Date(Date.UTC(y, m - 1, day + deltaDays));
  return utcDateKey(dt);
}

function resolveAsOf(entries: DailyEntry[], asOf?: string): string {
  return asOf ?? entries[0]?.date ?? utcDateKey();
}

/** First occurrence wins — pass recent-first so duplicates keep the newest. */
export function indexEntriesByDate(entries: DailyEntry[]): Map<string, DailyEntry> {
  const map = new Map<string, DailyEntry>();
  for (const e of entries) {
    if (e.date && !map.has(e.date)) map.set(e.date, e);
  }
  return map;
}

/**
 * `days` UTC calendar days ending on `asOf` (inclusive), most-recent first.
 * Missing check-ins appear as `{ date, entry: null }`.
 */
export function calendarWindow(
  entries: DailyEntry[],
  days: number,
  asOf?: string,
): CalendarDay[] {
  const end = resolveAsOf(entries, asOf);
  const byDate = indexEntriesByDate(entries);
  const out: CalendarDay[] = [];
  for (let i = 0; i < days; i++) {
    const date = shiftDateKey(end, -i);
    out.push({ date, entry: byDate.get(date) ?? null });
  }
  return out;
}

/**
 * Mean of present numeric values. Null, undefined, and missing days are
 * omitted — never coerced to 0.
 */
export function averagePresent(values: Array<number | null | undefined>): number | null {
  const nums = values.filter((v): v is number => typeof v === 'number' && Number.isFinite(v));
  if (nums.length === 0) return null;
  return nums.reduce((a, b) => a + b, 0) / nums.length;
}

/** Sum of training load over the last N calendar days. Missing days = 0. */
export function sumLoad(entries: DailyEntry[], days: number, asOf?: string): number {
  return calendarWindow(entries, days, asOf).reduce(
    (sum, day) => sum + (day.entry?.trainingLoad ?? 0),
    0,
  );
}

/**
 * ACWR = acute (7-day calendar sum) ÷ chronic (28-day calendar sum ÷ 4).
 * Gabbett (2016) sweet spot: 0.8–1.3 low, >1.5 danger zone (Section 18.1 /
 * 18.2 — cited with the caveat that this figure has been contested).
 *
 * Section 18.4 — these ACWR cutoffs are fixed constants. They are NOT
 * loaded from coaches/{uid}/teamSettings.
 */
export function calculateACWR(
  entries: DailyEntry[],
  asOf?: string,
): { acwr: number; acute7: number; chronicAvgWeekly: number } {
  const acute7 = sumLoad(entries, 7, asOf);
  const chronic28Sum = sumLoad(entries, 28, asOf);
  const chronicAvgWeekly = chronic28Sum / 4;

  // Guard against div-by-zero for new athletes with no chronic baseline.
  const acwr = chronicAvgWeekly > 0 ? acute7 / chronicAvgWeekly : 1.0;

  return { acwr, acute7, chronicAvgWeekly };
}

/**
 * Simplified Banister Fitness-Fatigue model. Fitness decays slowly
 * (~42 days), Fatigue decays quickly (~7 days). Performance = Fitness − Fatigue.
 *
 * Each UTC calendar day from the oldest check-in through `asOf` is one
 * decay step. Missing days contribute 0 load but still decay — otherwise a
 * 3-day gap would be treated as three consecutive sessions.
 * `entries` may be in either order.
 */
export function calculateFitnessFatigue(
  entries: DailyEntry[],
  asOf?: string,
): {
  fitness: number;
  fatigue: number;
  performanceIndex: number;
} {
  const fitnessTau = 42;
  const fatigueTau = 7;
  let fitness = 0;
  let fatigue = 0;

  if (entries.length === 0) {
    return { fitness, fatigue, performanceIndex: 0 };
  }

  const byDate = indexEntriesByDate(entries);
  const dates = [...byDate.keys()].sort();
  const start = dates[0];
  const end = asOf ?? dates[dates.length - 1];
  if (end < start) {
    return { fitness, fatigue, performanceIndex: 0 };
  }

  for (let key = start; key <= end; key = shiftDateKey(key, 1)) {
    const load = byDate.get(key)?.trainingLoad ?? 0;
    fitness = fitness * Math.exp(-1 / fitnessTau) + load;
    fatigue = fatigue * Math.exp(-1 / fatigueTau) + load;
  }

  return { fitness, fatigue, performanceIndex: fitness - fatigue };
}

/**
 * Recovery trend: HRV-based if the athlete's device provides it
 * (Tier 1). Otherwise falls back to a rolling baseline built from
 * resting heart rate, sleep, and self-reported fatigue (Section 9 —
 * "recovery trend degrades gracefully" rather than blocking).
 *
 * Windows are 7 calendar days vs the prior 7 calendar days. Wellness
 * metrics average only logged values (see averagePresent).
 */
export function calculateRecoveryTrend(
  entriesRecentFirst: DailyEntry[],
  asOf?: string,
): {
  trend: 'improving' | 'stable' | 'worsening';
  usedHRV: boolean;
} {
  const last7 = calendarWindow(entriesRecentFirst, 7, asOf);
  const prev7 = calendarWindow(entriesRecentFirst, 14, asOf).slice(7, 14);

  const hrvRecent = averagePresent(last7.map((d) => d.entry?.hrv));
  const hrvPrior = averagePresent(prev7.map((d) => d.entry?.hrv));

  if (hrvRecent != null && hrvPrior != null && hrvPrior !== 0) {
    const change = (hrvRecent - hrvPrior) / hrvPrior;
    if (change > 0.05) return { trend: 'improving', usedHRV: true };
    if (change < -0.05) return { trend: 'worsening', usedHRV: true };
    return { trend: 'stable', usedHRV: true };
  }

  let worseVotes = 0;
  let betterVotes = 0;

  const rhrRecent = averagePresent(last7.map((d) => d.entry?.restingHeartRate));
  const rhrPrior = averagePresent(prev7.map((d) => d.entry?.restingHeartRate));
  if (rhrRecent != null && rhrPrior != null && rhrPrior !== 0) {
    if (rhrRecent > rhrPrior * 1.03) worseVotes++;
    else if (rhrRecent < rhrPrior * 0.97) betterVotes++;
  }

  const sleepRecent = averagePresent(last7.map((d) => d.entry?.sleepHours));
  const sleepPrior = averagePresent(prev7.map((d) => d.entry?.sleepHours));
  if (sleepRecent != null && sleepPrior != null) {
    if (sleepRecent < sleepPrior - 0.5) worseVotes++;
    else if (sleepRecent > sleepPrior + 0.5) betterVotes++;
  }

  const fatigueRecent = averagePresent(last7.map((d) => d.entry?.fatigueScore));
  const fatiguePrior = averagePresent(prev7.map((d) => d.entry?.fatigueScore));
  if (fatigueRecent != null && fatiguePrior != null) {
    if (fatigueRecent > fatiguePrior + 0.4) worseVotes++;
    else if (fatigueRecent < fatiguePrior - 0.4) betterVotes++;
  }

  if (worseVotes > betterVotes) return { trend: 'worsening', usedHRV: false };
  if (betterVotes > worseVotes) return { trend: 'improving', usedHRV: false };
  return { trend: 'stable', usedHRV: false };
}

/**
 * Fatigue staying elevated for several days (Step 4).
 * True only when each of the last 4 calendar days has a check-in and
 * every fatigueScore is >= 4. A gap breaks the streak — we do not invent
 * a fatigue value for a missed day.
 */
export function isFatiguePersistent(entriesRecentFirst: DailyEntry[], asOf?: string): boolean {
  const last4 = calendarWindow(entriesRecentFirst, 4, asOf);
  if (last4.length < 4) return false;
  return last4.every((day) => day.entry != null && day.entry.fatigueScore >= 4);
}

File: MODEL/athleteiq-risk-engine/calculations.py
"""
Direct port of functions/src/calculations.ts — same formulas, same
thresholds, same variable names where possible, so behavior matches
exactly what your Cloud Function would have done.

Research-verified formulas (Section 18.1):
- Training Load = Duration x RPE (session-RPE method)
- ACWR = 7-day acute load / 28-day chronic average weekly load
- Fitness/Fatigue = Banister impulse-response model
"""
from __future__ import annotations
from dataclasses import dataclass
from typing import Optional
import math


@dataclass
class Entry:
    """One check-in, shaped exactly like the objects index.ts builds
    from Firestore docs before calling assessRisk()."""
    date: str
    training_load: Optional[float]
    sleep_hours: Optional[float]
    resting_heart_rate: Optional[float]
    hrv: Optional[float]
    fatigue_score: int


def sum_load(entries: list[Entry], days: int) -> float:
    """Sum of training load over the most recent N days (entries must
    already be sorted recent-first, matching index.ts's Firestore query)."""
    recent = entries[:days]
    return sum(e.training_load or 0 for e in recent)


def calculate_acwr(entries: list[Entry]) -> tuple[float, float, float]:
    """ACWR = acute (7-day sum) / chronic (28-day sum / 4). Gabbett (2016)
    sweet spot 0.8-1.3, danger zone >1.5 (Section 18.1/18.2, contested
    in the literature — always labeled as a signal, not a certainty)."""
    acute7 = sum_load(entries, 7)
    chronic28_sum = sum_load(entries, 28)
    chronic_avg_weekly = chronic28_sum / 4
    acwr = (acute7 / chronic_avg_weekly) if chronic_avg_weekly > 0 else 1.0
    return acwr, acute7, chronic_avg_weekly


def calculate_fitness_fatigue(entries_oldest_first: list[Entry]) -> tuple[float, float, float]:
    """Simplified Banister Fitness-Fatigue model. Fitness decays slowly
    (~42 day time constant), Fatigue decays quickly (~7 days).
    Performance = Fitness - Fatigue."""
    fitness_tau = 42
    fatigue_tau = 7
    fitness = 0.0
    fatigue = 0.0
    for e in entries_oldest_first:
        load = e.training_load or 0
        fitness = fitness * math.exp(-1 / fitness_tau) + load
        fatigue = fatigue * math.exp(-1 / fatigue_tau) + load
    return fitness, fatigue, fitness - fatigue


def _avg_of(nums: list[float]) -> float:
    return sum(nums) / len(nums) if nums else 0.0


def calculate_recovery_trend(entries_recent_first: list[Entry]) -> tuple[str, bool]:
    """HRV-based trend if the athlete's device provides it (Tier 1).
    Otherwise a rolling baseline from resting HR + sleep + fatigue
    (Section 9 — degrades gracefully rather than blocking)."""
    last7 = entries_recent_first[:7]
    prev7 = entries_recent_first[7:14]

    has_hrv = any(e.hrv is not None for e in last7) and any(e.hrv is not None for e in prev7)

    if has_hrv:
        recent = _avg_of([e.hrv for e in last7 if e.hrv is not None])
        prior = _avg_of([e.hrv for e in prev7 if e.hrv is not None])
        if prior == 0:
            return "stable", True
        change = (recent - prior) / prior
        if change > 0.05:
            return "improving", True
        if change < -0.05:
            return "worsening", True
        return "stable", True

    worse_votes = 0
    better_votes = 0

    rhr_recent = _avg_of([e.resting_heart_rate for e in last7 if e.resting_heart_rate is not None])
    rhr_prior = _avg_of([e.resting_heart_rate for e in prev7 if e.resting_heart_rate is not None])
    if rhr_recent and rhr_prior:
        if rhr_recent > rhr_prior * 1.03:
            worse_votes += 1
        elif rhr_recent < rhr_prior * 0.97:
            better_votes += 1

    sleep_recent = _avg_of([e.sleep_hours for e in last7 if e.sleep_hours is not None])
    sleep_prior = _avg_of([e.sleep_hours for e in prev7 if e.sleep_hours is not None])
    if sleep_recent and sleep_prior:
        if sleep_recent < sleep_prior - 0.5:
            worse_votes += 1
        elif sleep_recent > sleep_prior + 0.5:
            better_votes += 1

    fatigue_recent = _avg_of([e.fatigue_score for e in last7])
    fatigue_prior = _avg_of([e.fatigue_score for e in prev7])
    if fatigue_prior:
        if fatigue_recent > fatigue_prior + 0.4:
            worse_votes += 1
        elif fatigue_recent < fatigue_prior - 0.4:
            better_votes += 1

    if worse_votes > better_votes:
        return "worsening", False
    if better_votes > worse_votes:
        return "improving", False
    return "stable", False


def is_fatigue_persistent(entries_recent_first: list[Entry]) -> bool:
    """Is Fatigue staying elevated for several days instead of clearing? (Step 4)"""
    last4 = entries_recent_first[:4]
    if len(last4) < 4:
        return False
    return all(e.fatigue_score >= 4 for e in last4)
